

"""
    snakemake --keep-going -j 999999 --cluster "sbatch -A {cluster.account} --C {cluster.C} -c {cluster.cpus-per-task} -N {cluster.Nodes}  -t {cluster.runtime} -J {cluster.jobname} --mail-type={cluster.mail_type} --mail-user={cluster.mail}" --cluster-config cluster.json --configfile experiments.json --latency-wait 100 --verbose -n
"""

shell.prefix("set -o pipefail; ")
configfile: "experiments.json"

# wildcard_constraints:
#     nr_reads="[\d]+",

####################################################
########## standard python functions ###############
####################################################

import re
import os
import errno
import shutil
import glob



def mkdir_p(path):
    print("creating", path)
    try:
        os.makedirs(path)
    except OSError as exc:  # Python >2.5
        if exc.errno == errno.EEXIST and os.path.isdir(path):
            pass
        else:
            raise

rule all:
   input:  config["ROOT_OUT"] + "/evaluation/accuracy.result",
           config["ROOT_OUT"] + "/evaluation/time_and_mem.result"
            # expand(config["ROOT_OUT"] + "/evaluation/{tool}/{dataset}/{read_length}.result",  dataset =  ["SIM1", "SIM2", "SIM3"], read_length =  ["100", "150", "200", "250", "300"], tool=["minimap2", "bwa_mem", "strobealign"]) #"strobealign", 'bwa', 'bowtie', 'accelalign',


rule mason_simulate_variants:
    input:  ref = config["HG38"]
    output: sim_vcf =  config["ROOT_OUT"] + "/reads/{dataset}/variations.vcf",
    run:
        if wildcards.dataset == "SIM1":
            shell("mason_variator -ir {input.ref} -ov {output.sim_vcf}")
        elif wildcards.dataset == "SIM2":
            shell("mason_variator --sv-indel-rate 0.000001 --snp-rate 0.001 --small-indel-rate 0.00001 --max-small-indel-size 20  -ir {input.ref} -ov {output.sim_vcf}")
        elif wildcards.dataset == "SIM3":
            shell("mason_variator --sv-indel-rate 0.000005 --snp-rate 0.001 --small-indel-rate 0.0001 --max-small-indel-size 50   -ir {input.ref} -ov {output.sim_vcf}")
        elif wildcards.dataset == "SIM4":
            shell("mason_variator --sv-indel-rate 0.00001 --snp-rate 0.005 --small-indel-rate 0.0005 --max-small-indel-size 50   -ir {input.ref} -ov {output.sim_vcf}")


rule mason_simulate_reads:
    input:  ref = config["HG38"],
            sim_vcf = rules.mason_simulate_variants.output.sim_vcf
    output: sim_fq =  config["ROOT_OUT"] + "/reads/{dataset}/{read_length}.fq",
            sim_sam =  config["ROOT_OUT"] + "/reads/{dataset}/{read_length}.sam",
    run:

        if wildcards.read_length == "250" or wildcards.read_length == "300" or wildcards.read_length == "500": # to not give abort trap in simulator due to too low fragment size
            shell("mason_simulator -ir {input.ref} -n 10000000 -iv {input.sim_vcf} --illumina-read-length {wildcards.read_length} -o {output.sim_fq} -oa {output.sim_sam} --fragment-mean-size 700")
        else:
            shell("mason_simulator -ir {input.ref} -n 10000000 -iv {input.sim_vcf} --illumina-read-length {wildcards.read_length} -o {output.sim_fq} -oa {output.sim_sam}")



rule minimap2_index:
    input: ref = config["HG38"]
    output: mm_index =  config["HG38"] + ".mmi",
            time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/minimap2/indexing_time_and_mem.txt",
    run:
        shell("/usr/bin/time -v  minimap2 -t 1 -k21 -w11 -d {output.mm_index} {input.ref}  2>&1 | tee {output.time_and_mem}") #  2> {output.time_and_mem}

rule minimap2_align:
    input: fastq =   rules.mason_simulate_reads.output.sim_fq,
            # ref = config["HG38"],
            index = rules.minimap2_index.output.mm_index
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/minimap2/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/minimap2/{dataset}/{read_length}.sam"
    run:
        shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
        shell("echo Copied reads to: $SLURM_JOB_ID")
        tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
        tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
        tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
        shell("/usr/bin/time -v  minimap2 -o {tmp_sam} -t 16 -ax sr {input.index} {tmp_fq} 2>&1 | tee {tmp_time_and_mem}  ")
        shell("mv {tmp_time_and_mem} {output.time_and_mem}")
        shell("mv {tmp_sam} {output.sam}")

rule minimap2_map:
    input: fastq =   rules.mason_simulate_reads.output.sim_fq,
            # ref = config["HG38"],
            index = rules.minimap2_index.output.mm_index
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/minimap2_map/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/minimap2_map/{dataset}/{read_length}.sam"
    run:
        shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
        shell("echo Copied reads to: $SLURM_JOB_ID")
        tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
        tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
        tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
        shell("/usr/bin/time -v  minimap2 -o {tmp_sam} -t 16 -x sr {input.index} {tmp_fq} 2>&1 | tee {tmp_time_and_mem} ")
        shell("mv {tmp_time_and_mem} {output.time_and_mem}")
        shell("mv {tmp_sam} {output.sam}")


rule bwa_mem_index:
    input: ref = config["HG38"]
    output: bwa_index =  config["HG38"] + ".bwt",
            time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/bwa_mem/indexing_time_and_mem.txt",
    run:
        shell("/usr/bin/time -v  bwa index {input.ref}  2>&1 | tee {output.time_and_mem}") #  2> {output.time_and_mem}

rule bwa_mem_align:
    input: fastq =   rules.mason_simulate_reads.output.sim_fq,
            index = config["HG38"] + ".bwt",
            ref = config["HG38"]
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/bwa_mem/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/bwa_mem/{dataset}/{read_length}.sam"
    run:
        shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
        shell("echo Copied reads to: $SLURM_JOB_ID")
        tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
        tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
        tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
        shell("/usr/bin/time -v  bwa mem -o {tmp_sam}  -t 16  {input.ref} {tmp_fq} 2>&1 | tee {tmp_time_and_mem} ")
        shell("mv {tmp_time_and_mem} {output.time_and_mem}")
        shell("mv {tmp_sam} {output.sam}")


rule bwa_mem2_index:
    input: ref = config["HG38"]
    output: bwa_mem2_index = config["HG38"] + ".bwa_mem2.ann", # just one of the files as marker for completion
            time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/bwa_mem2/indexing_time_and_mem.txt",
    run:
        idxbase = config["HG38"] + ".bwa_mem2"
        shell("/usr/bin/time -v  bwa-mem2 index -p {idxbase} {input.ref}  2>&1 | tee {output.time_and_mem}") #  2> {output.time_and_mem}

rule bwa_mem2_align:
    input:  fastq = rules.mason_simulate_reads.output.sim_fq,
            index = config["HG38"] + ".bwa_mem2.ann",
            ref = config["HG38"]
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/bwa_mem2/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/bwa_mem2/{dataset}/{read_length}.sam"
    run:
        shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
        shell("echo Copied reads to: $SLURM_JOB_ID")
        tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
        tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
        tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
        idxbase = config["HG38"] + ".bwa_mem2"
        shell("/usr/bin/time -v  bwa-mem2 mem -o {tmp_sam} -t 16 {idxbase} {tmp_fq} 2>&1 | tee {tmp_time_and_mem} ")
        shell("mv {tmp_time_and_mem} {output.time_and_mem}")
        shell("mv {tmp_sam} {output.sam}")


rule strobealign:
    input:  ref = config["HG38"],
            fastq =   rules.mason_simulate_reads.output.sim_fq
            # index = rules.minimap2_index.output.mm_index
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/strobealign/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/strobealign/{dataset}/{read_length}.sam"
    run: 
        shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
        shell("echo Copied reads to: $SLURM_JOB_ID")
        tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
        tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
        tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
        shell("/usr/bin/time -v  strobealign -t 16 -r {wildcards.read_length} -o {tmp_sam} {input.ref} {tmp_fq} 2>&1 | tee {tmp_time_and_mem}") #  
        shell("mv {tmp_time_and_mem} {output.time_and_mem}")
        shell("mv {tmp_sam} {output.sam}")


rule strobealign_map:
    input:  ref = config["HG38"],
            fastq =   rules.mason_simulate_reads.output.sim_fq
            # index = rules.minimap2_index.output.mm_index
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/strobealign_map/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/strobealign_map/{dataset}/{read_length}.sam"
    run: 
        shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
        shell("echo Copied reads to: $SLURM_JOB_ID")
        tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
        tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
        tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
        shell("/usr/bin/time -v  strobealign -t 16 -r {wildcards.read_length} -x -o {tmp_sam} {input.ref} {tmp_fq} 2>&1 | tee {tmp_time_and_mem}") #  
        shell("mv {tmp_time_and_mem} {output.time_and_mem}")
        shell("mv {tmp_sam} {output.sam}")

rule strobealign_mixed:
    input:  ref = config["HG38"],
            fastq =   rules.mason_simulate_reads.output.sim_fq
            # index = rules.minimap2_index.output.mm_index
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/strobealign_mixed/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/time_and_mem/strobealign_mixed/{dataset}/{read_length}.sam"
    run: 
        shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
        shell("echo Copied reads to: $SLURM_JOB_ID")
        tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
        tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
        tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
        shell("/usr/bin/time -v  strobealign-mixed -t 16 -r {wildcards.read_length} -x -o {tmp_sam} {input.ref} {tmp_fq} 2>&1 | tee {tmp_time_and_mem}") #  
        shell("mv {tmp_time_and_mem} {output.time_and_mem}")
        shell("mv {tmp_sam} {output.sam}")

rule accelalign_index:
    input: ref = config["HG38"]
    output: accelalign_index =  config["HG38"] + ".hash",
            time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/accelalign/indexing_time_and_mem.txt"
    run:
        shell("/usr/bin/time -v  accindex -t 1 {input.ref} 2>&1 | tee {output.time_and_mem}") #  2> {output.time_and_mem}

rule accelalign_align:
    input: fastq =   rules.mason_simulate_reads.output.sim_fq,
            index = rules.accelalign_index.output.accelalign_index,
            # index = rules.minimap2_index.output.mm_index
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/accelalign/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/accelalign/{dataset}/{read_length}.sam"
    run:
        ref = config["HG38"]
        shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
        shell("echo Copied reads to: $SLURM_JOB_ID")
        tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
        tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
        tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
        shell("/usr/bin/time -v  accalign -t 16 -o {tmp_sam} {ref} {tmp_fq} 2>&1 | tee {tmp_time_and_mem}")
        shell("mv {tmp_time_and_mem} {output.time_and_mem}")
        shell("mv {tmp_sam} {output.sam}")


rule accelalign_map:
    input: fastq =   rules.mason_simulate_reads.output.sim_fq,
            index = rules.accelalign_index.output.accelalign_index,
            # index = rules.minimap2_index.output.mm_index
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/accelalign_map/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/accelalign_map/{dataset}/{read_length}.sam"
    run:
        ref = config["HG38"]
        shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
        shell("echo Copied reads to: $SLURM_JOB_ID")
        tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
        tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
        tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
        shell("/usr/bin/time -v  accalign -t 16 -x -o {tmp_sam} {ref} {tmp_fq} 2>&1 | tee {tmp_time_and_mem}")
        shell("mv {tmp_time_and_mem} {output.time_and_mem}")
        shell("mv {tmp_sam} {output.sam}")


rule bowtie2_index:
    input: ref = config["HG38"]
    output: index =  config["HG38"] + "-bt2.1.bt2",
            time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/bowtie2/indexing_time_and_mem.txt"
    run:
        index =  config["HG38"] + "-bt2"
        shell("/usr/bin/time -v  bowtie2-build --threads 1 {input.ref} {index}  2>&1 | tee {output.time_and_mem}") #  2> {output.time_and_mem}

rule bowtie2_align:
    input: fastq =   rules.mason_simulate_reads.output.sim_fq,
            bt2_index =  config["HG38"] + "-bt2.1.bt2",
            # index = rules.minimap2_index.output.mm_index
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/bowtie2/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/bowtie2/{dataset}/{read_length}.sam"
    run:
        ref = config["HG38"]
        bt2_index = config["HG38"] + "-bt2"
        shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
        shell("echo Copied reads to: $SLURM_JOB_ID")
        tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
        tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
        tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"

        # if wildcards.read_length == "250" or wildcards.read_length == "300": 
        shell("/usr/bin/time -v  bowtie2 -p 16 -x {bt2_index} -U {tmp_fq}  -S {tmp_sam}  2>&1 | tee {tmp_time_and_mem}")
        shell("mv {tmp_time_and_mem} {output.time_and_mem}")
        shell("mv {tmp_sam} {output.sam}")


rule urmap_index:
    input: ref = config["HG38"]
    output: ufi_index =  config["HG38"] + ".ufi",
            time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/urmap/indexing_time_and_mem.txt",
    run:
        # urmap -make_ufi hg38.fa -veryfast -output hg38.ufi
        shell("/usr/bin/time -v  urmap -make_ufi {input.ref}  -output {output.ufi_index} 2>&1 | tee {output.time_and_mem}") 

rule urmap_align:
    input:  fastq =   rules.mason_simulate_reads.output.sim_fq,
            index = rules.urmap_index.output.ufi_index
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/urmap/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/urmap/{dataset}/{read_length}.sam"
    run:
        if wildcards.read_length == "300": # cannot index rye, gives error. and cannot align the 300nt reads
            shell("touch {output.time_and_mem}") 
            shell('> {output.sam}')
        else:
            mkdir_p(config["ROOT_OUT"] + "/alignments_PE/urmap/{0}/".format(wildcards.dataset))
            shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
            shell("echo Copied reads to: $SLURM_JOB_ID")
            ref = config["HG38"]
            tmp_index = "/scratch/$SLURM_JOB_ID/urmap_index.ufi"
            shell("mkdir -p /scratch/$SLURM_JOB_ID/")
            shell("/usr/bin/time -v  urmap -make_ufi {ref} -output {tmp_index}") 
            tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
            tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
            tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
            # tmp_index = "/scratch/$SLURM_JOB_ID/index.ufi"
            # urmap -map2 sample_R1.fastq.gz -reverse sample_R2.fastq.gz -ufi hg38.ufi -samout sample.sam -o {tmp_sam} -t 1 -ax sr
            shell("/usr/bin/time -v  urmap -map {tmp_fq} -ufi {tmp_index} -samout {tmp_sam} -threads 1 2>&1 | tee {tmp_time_and_mem}  ")
            shell("mv {tmp_time_and_mem} {output.time_and_mem}")
            shell("mv {tmp_sam} {output.sam}")


rule snap_index:
    input: ref = config["HG38"]
    output: snp_index =  config["HG38"] + "_snap_index/GenomeIndexHash",
            time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/snap/indexing_time_and_mem.txt",
    run:
        # snap
        shell("/usr/bin/time -v  snap-aligner index {input.ref}  {output.snp_index} 2>&1 | tee {output.time_and_mem}") 

rule snap_align:
    input:  fastq =   rules.mason_simulate_reads.output.sim_fq,
            index = rules.snap_index.output.snp_index
    output: time_and_mem =  config["ROOT_OUT"] + "/time_and_mem/snap/{dataset}/{read_length}/aligning_time_and_mem.txt",
            sam =  config["ROOT_OUT"] + "/alignments/snap/{dataset}/{read_length}.sam"
    run:
        if wildcards.read_length == "500": # cannot align longer than 400nt reads
            shell("touch {output.time_and_mem}") 
            shell('> {output.sam}')
        else:
            shell("cp {0} /scratch/$SLURM_JOB_ID/reads.fq".format(input.fastq))
            shell("echo Copied reads to: $SLURM_JOB_ID")
            ref = config["HG38"]
            tmp_index = "/scratch/$SLURM_JOB_ID/snap_index"
            shell("mkdir -p /scratch/$SLURM_JOB_ID/")
            shell("/usr/bin/time -v  snap-aligner index {ref}  {tmp_index} -t1") 
            shell("echo Copied index to: $SLURM_JOB_ID")
            tmp_time_and_mem = "/scratch/$SLURM_JOB_ID/aligning_time_and_mem.txt"
            tmp_sam = "/scratch/$SLURM_JOB_ID/aligned.sam"
            tmp_fq = "/scratch/$SLURM_JOB_ID/reads.fq"
            # snap-aligner paired index-dir read1.fq read2.fq -o output.sam
            # tmp_index = "/scratch/$SLURM_JOB_ID/snap_index/"
            # tmp_index = config["HG38"] + "_snap_index"
            shell("/usr/bin/time -v  snap-aligner single {tmp_index} {tmp_fq} -o {tmp_sam} -t 16  2>&1 | tee {tmp_time_and_mem}  ")
            shell("mv {tmp_time_and_mem} {output.time_and_mem}")
            shell("mv {tmp_sam} {output.sam}")


rule accuracy:
    input: aln =  config["ROOT_OUT"] + "/alignments/{tool}/{dataset}/{read_length}.sam",
            truth =  config["ROOT_OUT"] + "/reads/{dataset}/{read_length}.sam"
    output: out = config["ROOT_OUT"] + "/evaluation/{tool}/{dataset}/{read_length}.result"
    run:
        if wildcards.tool == "urmap" and wildcards.read_length == "300":
            shell('echo "-,-,-" > {output.out}')
        elif wildcards.tool == "snap" and wildcards.read_length == "500":
            shell('echo "-,-,-" > {output.out}')
        elif wildcards.tool == "strobealign_mixed" and wildcards.read_length == "500":
            shell('echo "-,-,-" > {output.out}')
        elif wildcards.tool == "minimap2_map" or wildcards.tool == "strobealign_map":
            shell("/usr/bin/time -v  python get_accuracy.py --truth {input.truth} --predicted_paf  {input.aln} > {output.out} ")
        else:
            shell("/usr/bin/time -v  python get_accuracy.py --truth {input.truth} --predicted_sam  {input.aln} > {output.out} ")



rule summary_alignment_stats:
    input: accuracy_files = expand(config["ROOT_OUT"] + "/evaluation/{tool}/{dataset}/{read_length}.result",  dataset =  ["SIM1", "SIM2", "SIM3", "SIM4"], read_length =  ["50", "75","100", "150", "200", "250", "300", "500"], tool=["minimap2", "bwa_mem", "strobealign",'accelalign', "bowtie2", "strobealign_map", "minimap2_map", "accelalign_map", 'snap', 'urmap',  "bwa_mem2"]) #"strobealign_mixed",
    output: out = config["ROOT_OUT"] + "/evaluation/accuracy.result"
    run:
        shell('echo "tool,dataset,read_length,aligned,accuracy,overaligned" > {output.out}')
        for f in input.accuracy_files:
            string = f.split("/")
            tool = string[-3]
            dataset = string[-2]
            read_length = string[-1].split(".")[0]
            shell("echo -n {tool},{dataset},{read_length}, >> {output.out} ")
            shell("cat {f} >> {output.out}")
            


rule summary_time_and_mem:
    input: time_and_mem_files =  expand(config["ROOT_OUT"] + "/time_and_mem/{tool}/{dataset}/{read_length}/aligning_time_and_mem.txt",  dataset =  ["SIM1", "SIM2", "SIM3", "SIM4"], read_length =  ["50", "75","100", "150", "200", "250", "300", "500"], tool=["minimap2", "bwa_mem", "strobealign", 'accelalign', "bowtie2", "strobealign_map", "minimap2_map", "accelalign_map", 'snap', 'urmap',  "bwa_mem2"]) #"strobealign_mixed",
    output: out = config["ROOT_OUT"] + "/evaluation/time_and_mem.result"
    run:
        # summary_file = open(out, "w")
        shell('echo "tool,dataset,read_length,time,memory" > {output.out}')

        for f in input.time_and_mem_files:
            string = f.split("/")
            tool = string[-4]
            dataset = string[-3]
            read_length = string[-2]
            shell("/usr/bin/time -v  python summarize_time_and_memory.py {tool} {dataset} {read_length} {f}  >> {output.out} ")









